### PR TITLE
68 arbitrary meter data issuance

### LINKED ISSUE #68: Add endpoint for arbitrary metering issuance
## Description

Add an endpoint that allows users to pass a serialised CSV of metering data to issue against a specified device and accont. This will live alongside the registry-wide issue by date range function.

## Dependencies

- with third parties, data etc.

## Tests

- [] What tests are needed to accomplish the success criteria

## Success Criteria

- [] Checkboxes describing every step needed to be taken for defining this ticket as Done.

### LINKED ISSUE #73: Change max device bundle end ID validation to work with bulk GC bundle inserts
## Description

Currently, the test in `validate_granular_certificate_bundle` that checks if the `bundle_range_id_start` is equal to 1 plus the highest `bundle_range_id_end` for that device relies on the last GC bundle being in the database. This approach won't work when bulk-writing bundles to the database as the previous GC bundle won't be in the database at the time the present one is checked. Given it appears to take >15minutes to issue a month's worth of GCs for a single device, we need to be using bulk inserts, therefore this part of the validation will need refactoring. 

## Dependencies

- with third parties, data etc.

## Tests

- [] What tests are needed to accomplish the success criteria

## Success Criteria

- [] Checkboxes describing every step needed to be taken for defining this ticket as Done.

### PR DESCRIPTION
**Pull request by CAGalbraith**
## Description

Merge #72 first, this branch builds on that feature.

- Closes #68 - Adds a new endpoint `measurement/submit_readings` that accepts a serialised CSV of arbitrary meter readings associated with a set of devices that must already be present on the database. Includes a new meter client `ManualSubmissionMeterClient` that parses and issues GC Bundles for these meter records, saved to the DB using the `MeasurementReport` class.
- Closes #73 - `issue_certificates_by_device_in_date_range` has been tweaked to not check the DB for the device's max bundle range end ID each time it iterates through the list of GC bundles, in order to make batch issuing possible (issuing one at a time for a month causes the container to sieze up)

## Has This Been Tested?

- [x] Yes

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration_

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: gc_registry/measurement/routes.py [hunk 1]
```diff
@@ -1,17 +1,61 @@
...
+from gc_registry.measurement.services import parse_measurement_json
 
 # Router initialisation
 router = APIRouter(tags=["Measurements"])
 
 ### Device Meter Readings ###
 
 
+@router.post("/submit_readings", response_model=models.MeasurementSubmissionResponse)
+def submit_readings(
```
Comment (ClemAtt) [inline gc_registry/measurement/routes.py:17]: What does providing the `measurement_json` to the front-end look like for the UI? It looks like in FastAPI you might be able to skip some of this with something like the following:

```
@app.post("/upload")
def upload(file: UploadFile = File(...)):
    csvReader = csv.DictReader(codecs.iterdecode(file.file, 'utf-8'))
    data = {}
    for rows in csvReader:             
        key = rows['Id']  # Assuming a column named 'Id' to be the primary key
        data[key] = rows  
    
    file.file.close()
    return data

```

File: gc_registry/measurement/routes.py [hunk 2]
```diff
@@ -1,17 +1,61 @@
...
+
+    Args:
+        measurement_json (str): A JSON-serialised CSV file containing the meter readings.
+
+    Returns:
+        models.MeasurementSubmissionResponse: A summary of the readings submitted.
+    """
+    measurement_df = parse_measurement_json(measurement_json, to_df=True)
+
+    readings = models.MeasurementReport.create(
```
Comment (ClemAtt) [inline gc_registry/measurement/routes.py:35]: Is this creation step necessary. If the registry stores duplicates of certificates in measurement data isn't that a bit redundant?

Can the incoming data just be read, validated and written to GCBs or respond with an Exception

File: gc_registry/tests/measurement/test_measurement_routes.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    return """
+    [
+        {"device_id": "device_1", "interval_usage": 10, "interval_start_datetime": "2024-11-18T10:00:00", "interval_end_datetime": "2024-11-18T11:00:00"},
+        {"device_id": "device_1", "interval_usage": 15, "interval_start_datetime": "2024-11-18T11:00:00", "interval_end_datetime": "2024-11-18T12:00:00"},
+        {"device_id": "device_2", "interval_usage": 20, "interval_start_datetime": "2024-11-18T10:00:00", "interval_end_datetime": "2024-11-18T11:00:00"}
+    ]
+    """
+
+
+def test_submit_readings_success(api_client, db_write_session, db_read_session, esdb_client, valid_measurement_json, parsed_measurement_df, monkeypatch
```
Comment (ClemAtt) [inline gc_registry/tests/measurement/test_measurement_routes.py:20]: Had a go at adding a test, but it's complaining about the `measurement_json` being missing regardless of whether I use `json`, `data`, `params` arguments. Think maybe it needs to be in the request body

File: gc_registry/measurement/routes.py [hunk 4]
```diff
@@ -1,17 +1,61 @@
...
 # Router initialisation
 router = APIRouter(tags=["Measurements"])
 
 ### Device Meter Readings ###
 
 
+@router.post("/submit_readings", response_model=models.MeasurementSubmissionResponse)
+def submit_readings(
+    measurement_json: str,
+    write_session: Session = Depends(db.get_write_session),
```
Comment (ClemAtt) [inline gc_registry/measurement/routes.py:19]: Should also take the `device_id` in order to be validated

File: gc_registry/measurement/routes.py (hunk 1 above)
Comment (CAGalbraith) [reply gc_registry/measurement/routes.py:17]: Yeah I saw that bit of code as well, for the front end that will likely be the approach but for now it's just for the demo notebook. It'll be provide a path to the CSV to the serialize_meter_reasings function that can then just be passed directly as `content` to the endpoint

File: gc_registry/measurement/routes.py (hunk 2 above)
Comment (CAGalbraith) [reply gc_registry/measurement/routes.py:35]: I mean the registry needs to keep the measurement data, even if it doesn't result in a GC issuance right?

Comment (CAGalbraith) [reply gc_registry/measurement/routes.py:35]: Will ask M-RETS but preference currently to not retain measurement records, any improperly issued GC rectification would need the device to send their records in again anyway. If keeping, consider retention period.

**Review by ClemAtt — APPROVED** [review]

**Pull request merged**